biz.aQute.resolve/src/biz/aQute/resolve/Bndrun.java

```diff
@@ -0,0 +1,146 @@
+			",\\\n\t", new HeaderClauseFormatter(), null);
+
+	public Bndrun(Workspace workspace, File propertiesFile) throws Exception {
+		super(workspace, propertiesFile);
+	}
+
+	public void export(File targetDir) throws Exception {
+		try (Jar jar = getProjectLauncher().executable()) {
+			targetDir.mkdirs();
+			File jarFile = new File(targetDir, getNamePart(getPropertiesFile()) + ".jar");
+			jar.write(jarFile);
+		}
+	}
+
+	public String resolve(boolean failOnChanges, boolean writeOnChanges) throws Exception {
+		try (ProjectResolver projectResolver = new ProjectResolver(this)) {
+			Map<Resource,List<Wire>> resolution = projectResolver.resolve();
+			Set<Resource> resources = resolution.keySet();
+			List<VersionedClause> runBundles = new ArrayList<>();
+			for (Resource resource : resources) {
+				VersionedClause runBundle = ResourceUtils.toVersionClause(resource, "[===,==+)");
+				if (!runBundles.contains(runBundle)) {
+					runBundles.add(runBundle);
+				}
+			}
+			Collections.sort(runBundles, new Comparator<VersionedClause>() {
+				@Override
+				public int compare(VersionedClause a, VersionedClause b) {
+					int diff = a.getName().compareTo(b.getName());
+					return (diff != 0) ? diff : a.getVersionRange().compareTo(b.getVersionRange());
+				}
+			});
+
+			File runFile = getPropertiesFile();
+			BndEditModel bem = new BndEditModel(getWorkspace());
+			Document doc = new Document(IO.collect(runFile));
+			bem.loadFrom(doc);
+
+			List<VersionedClause> bemRunBundles = bem.getRunBundles() != null ? bem.getRunBundles()
+					: new ArrayList<VersionedClause>();
+
+			String originalRunbundlesString = headerClauseListFormatter.convert(bemRunBundles);
+			logger.debug("Original -runbundles was:\n\t {}", originalRunbundlesString);
+			String runbundlesString = headerClauseListFormatter.convert(runBundles);
+			logger.debug("Resolved -runbundles is:\n\t {}", runbundlesString);
+
+			List<VersionedClause> deltaAdd = new ArrayList<>(runBundles);
+			deltaAdd.removeAll(bemRunBundles);
+			List<VersionedClause> deltaRemove = new ArrayList<>(bemRunBundles);
+			deltaRemove.removeAll(runBundles);
+			boolean added = bemRunBundles.addAll(deltaAdd);
+			boolean removed = bemRunBundles.removeAll(deltaRemove);
+			if (added || removed) {
+				if (failOnChanges && !bemRunBundles.isEmpty()) {
+					String message = String.format("The runbundles have changed. Failing the build!\nWas: %s\nIs: %s",
+							originalRunbundlesString, runbundlesString);
+					throw new Exception(message);
+				}
+				bem.setRunBundles(bemRunBundles);
+				logger.info("Writing changes to {}", runFile.getAbsolutePath());
+				logger.info("{}:{}", Constants.RUNBUNDLES, bem.getDocumentChanges().get(Constants.RUNBUNDLES));
+				bem.saveChangesTo(doc);
+				if (writeOnChanges) {
+					IO.store(doc.get(), runFile);
+				}
+			}
+			return headerClauseListFormatter.convert(bem.getRunBundles());
```

Comment:
There must be a way to set the runbundles computed by  the resolver into the Run object so that test and export will use that value instead of the value in the bndrun file (which may not even be present).